### PR TITLE
Add better dask handling to DynamicAreaDefinitions

### LINKED ISSUE #344: Improve handling of dask arrays in DynamicAreaDefinition
#### Problem description

The `DynamicAreaDefinition`'s `freeze` method takes a series of longitudes and latitudes (or a geometry definition) and uses them to compute the outer bounds (extents) of the resulting AreaDefinition. This works for dask arrays as-is, but only because it currently casts the data to numpy arrays. The resulting computations could benefit (slightly) by allowing dask to compute things in parallel.

The current code that I've written in #342 includes two sets of min/max operations: an initial one and then another if the data crosses the anti-meridian. I'm not sure if these can be combined into a single set of map_blocks functions, but at a minimum the projection transformation and the initial min/max calculations could all be computed at the same time.

Here is the current version of my code in #342:

```
        xarr, yarr = proj4(np.asarray(lons), np.asarray(lats))
        xarr[xarr > 9e29] = np.nan
        yarr[yarr > 9e29] = np.nan
        xmin = np.nanmin(xarr)
        xmax = np.nanmax(xarr)
        ymin = np.nanmin(yarr)
        ymax = np.nanmax(yarr)
```

This could be split into three separate dask operations:

1. map_blocks version of pyproj transformation
2. `da.where` to get rid of infinity values and replace them with NaNs
3. Calculate nanmin/nanmax

Then compute the min/max values all in the same `da.compute` call.

### PR DESCRIPTION
**Pull request by djhoese**
See #344 for the details.

 - [x] Closes #344  
 - [x] Tests added 
 - [x] Tests passed 
 - [ ] Passes ``git diff origin/master **/*py | flake8 --diff`` 
 - [ ] Fully documented

### CODE REVIEW

**Comment by djhoese** [comment]
@mraspaud Well it works, but I'm not really happy with how gross the code got.

File: pyresample/test/test_geometry.py [hunk 1]
```diff
@@ -33,10 +33,17 @@
...
 import pyproj
 
+from pyproj import CRS
+
 try:
-    from pyproj import CRS
+    import dask.array as da
 except ImportError:
-    CRS = None
+    da = None
```
Comment (mraspaud) [inline pyresample/test/test_geometry.py:41]: Aren't da and xr requirements for tests?

File: pyresample/utils/proj4.py [hunk 2]
```diff
@@ -91,3 +92,33 @@ def get_geostationary_height(geos_area_crs):
...
+    crs = CRS.from_user_input(crs)
+    if not inverse:
+        crs_to = crs
+        crs_from = CRS(4326)
+    else:
+        crs_to = CRS(4326)
+        crs_from = crs
+    # CRS objects aren't thread-safe until pyproj 3.1+
+    # convert to WKT strings to be safe
+    result = da.map_blocks(_transform_dask_chunk, x, y,
```
Comment (mraspaud) [inline pyresample/utils/proj4.py:118]: Could you create the transformer here and pass transformer.transform to map_blocks? Or is it not thread safe either?

**Review by mraspaud — COMMENTED** [review]
Some questions to start with

File: pyresample/test/test_geometry.py (hunk 1 above)
Comment (djhoese) [reply pyresample/test/test_geometry.py:41]: Yes, I think we did this back when they weren't. I can remove them.

Comment (djhoese) [reply pyresample/test/test_geometry.py]: Fixed.

File: pyresample/utils/proj4.py (hunk 2 above)
Comment (djhoese) [reply pyresample/utils/proj4.py]: As discussed on slack, Transformer (and CRS and Proj) instances are not threadsafe until pyproj 3.1+. I've created a DaskFriendlyTransformer class that wraps this logic and mimics the pyproj interfaces. Hm, although now that I think about it, maybe I shouldn't have made the class *do* the actual transforming with dask but instead made a class that could be passed to dask?

File: pyresample/utils/proj4.py [hunk 3]
```diff
@@ -94,31 +94,48 @@ def get_geostationary_height(geos_area_crs):
...
+        self.src_crs = src_crs
+        self.dst_crs = dst_crs
+        self.kwargs = kwargs
+
+    @classmethod
+    def from_crs(cls, crs_from, crs_to, **kwargs):
+        """Create transformer object from two CRS objects."""
+        return cls(crs_from, crs_to, **kwargs)
+
+    def transform(self, x, y, inverse=False):
```
Comment (mraspaud) [inline pyresample/utils/proj4.py:123]: To be completely transparent, should this method take kwargs and pass them to pyproj's transform method?

**Review by mraspaud — APPROVED** [review]
LGTM

File: pyresample/utils/proj4.py (hunk 3 above)
Comment (djhoese) [reply pyresample/utils/proj4.py:123]: Ah good point. I'll fix that.

**Pull request merged**